# Lab 2

## Part A

A Calculating the probability of rolling m dice, each with n sides, and having the sum of the dice
be greater than x. This calculation should only be implemented for m = 1 through 4 and n is
at most 20.

In [1]:
import itertools
import pandas as pd
import numpy as np

In [2]:
# M: Number of dice
# N: How many sides each dice has
# X: Sum of dice to be greater than
def cal_greater_sum(m, n, x):
  # List out all possible rolls
  rolls = list(itertools.product(range(1, n + 1), repeat=m))
  # Find the sum of all possible rolls
  sums = np.array([sum(roll) for roll in rolls])
  # Find the proportion of those sums that are greater than or equal to the target x
  return np.mean(sums >= x)

## Part B

Determining the probability of rolling m dice, each with n sides, and having at least one die
show a value greater than or equal to x.

In [3]:
# M: Number of dice
# N: How many sides each dice has
# X: At least one dice to be greater than or equal to
def cal_at_least_one(m, n, x):
    # List out all possible rolls
    rolls = list(itertools.product(range(1, n + 1), repeat=m))
    # Count the number of rolls that have at least one dice greater than or equal to the target x
    count = len([roll for roll in rolls if max(roll) >= x])
    # Find the proportion of those rolls that have at least one dice greater than or equal to the target x
    return count / len(rolls)
    

## Evaluation

Evaluate the performance of your tool using the following scenario:
A player is in a challenging situation where they have a choice. They can choose between the
following challenges:
- Roll 1d20 and get a result of 17 or more
- Roll 3d4 and get a result of 10 or more
- Roll 4d6 and have at least one dice show a 6

Use your tool to compute the probabilities and report the best choice for the player.

In [4]:
# Just testing
print(cal_greater_sum(1, 20, 17))
print(cal_greater_sum(3, 4, 10))
print(cal_at_least_one(4, 6, 6))

0.2
0.15625
0.5177469135802469


The best choice for the player is to Roll a 4d6 and have at least one dice show a 6. There is a 51.77% chance of success with this choice.